### PR TITLE
[CLOB-1058] Allow FOK/IOC Reduce Only Orders

### PR DESCRIPTION
**Pull request by jonfung-dydx**
2 very small edits to allow FOK IOC reduce only errors, add validation for no replacement RO orders. E2e tests as well

### CODE REVIEW

File: protocol/x/clob/e2e/reduce_only_orders_test.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+	"github.com/cometbft/cometbft/types"
+	"github.com/dydxprotocol/v4-chain/protocol/dtypes"
+	testapp "github.com/dydxprotocol/v4-chain/protocol/testutil/app"
+	"github.com/dydxprotocol/v4-chain/protocol/testutil/constants"
+	clobtypes "github.com/dydxprotocol/v4-chain/protocol/x/clob/types"
+	satypes "github.com/dydxprotocol/v4-chain/protocol/x/subaccounts/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReduceOnlyOrders(t *testing.T) {
```
Comment (jakob-dydx) [inline protocol/x/clob/e2e/reduce_only_orders_test.go:15]: can we add tests for FOK as well?

**Review by jakob-dydx — APPROVED** [review]

File: protocol/x/clob/memclob/memclob.go [hunk 2]
```diff
@@ -1352,11 +1352,17 @@ func (m *MemClobPriceTimePriority) validateNewOrder(
...
+	// If the order is a reduce-only order, we should ensure the full size of the order does not increase the
+	// subaccount's current position size. Note that we intentionally do not validate that the reduce-only order
 	// does not change the subaccount's position _side_, and that will be validated if the order is matched.
 	// TODO(DEC-1228): use `MustValidateReduceOnlyOrder` and move this to `PerformStatefulOrderValidation`.
 	if order.IsReduceOnly() {
+		// If there exists an non-reduce only order with the same order id
+		// cancel the incoming reduce only order by returning an error.
+		if restingOrderExists && !existingRestingOrder.IsReduceOnly() {
+			return types.ErrReduceOnlyOrderReplacement
+		}
```
Comment (lucas-dydx) [inline protocol/x/clob/memclob/memclob.go:1364]: why is this logic necessary?

File: protocol/x/clob/e2e/reduce_only_orders_test.go [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+			for _, subaccount := range tc.expectedSubaccounts {
+				actualSubaccount := tApp.App.SubaccountsKeeper.GetSubaccount(ctx, *subaccount.Id)
+				require.Equal(t, subaccount, actualSubaccount)
+			}
+		})
+	}
+}
+
+func TestReduceOnlyOrderFailure(t *testing.T) {
```
Comment (lucas-dydx) [inline protocol/x/clob/e2e/reduce_only_orders_test.go:289]: can we add conditional order tests + long-term order tests for reduce-only?

File: protocol/x/clob/types/message_place_order_test.go [hunk 4]
```diff
@@ -250,6 +251,66 @@ func TestMsgPlaceOrder_ValidateBasic(t *testing.T) {
 					ReduceOnly:   true,
 				},
 			},
+		},
+		"short-term IOC reduce-only success": {
```
Comment (lucas-dydx) [inline protocol/x/clob/types/message_place_order_test.go:255]: add tests for conditional orders?

**Review by lucas-dydx — APPROVED** [review]

File: protocol/x/clob/e2e/reduce_only_orders_test.go (hunk 3 above)
Comment (jonfung-dydx) [reply protocol/x/clob/e2e/reduce_only_orders_test.go:289]: Future PR

File: protocol/x/clob/types/message_place_order_test.go (hunk 4 above)
Comment (jonfung-dydx) [reply protocol/x/clob/types/message_place_order_test.go:255]: future PR

File: protocol/x/clob/memclob/memclob.go (hunk 2 above)
Comment (jonfung-dydx) [reply protocol/x/clob/memclob/memclob.go:1364]: Started discussion in #p-reduce-only-ioc-fok

File: protocol/x/clob/e2e/reduce_only_orders_test.go (hunk 1 above)
Comment (jonfung-dydx) [reply protocol/x/clob/e2e/reduce_only_orders_test.go:15]: future PR

**Pull request merged**